rockcraft/models/project.py

```diff
@@ -326,30 +300,34 @@ def _validate_license(
             )
         return str(lic.id)  # type: ignore[reportUnknownMemberType]
 
-    @pydantic.validator("title", always=True)
+    @pydantic.model_validator(mode="before")
     @classmethod
-    def _validate_title(cls, title: str | None, values: Mapping[str, Any]) -> str:
+    def _validate_title(cls, values: dict[str, Any]) -> dict[str, Any]:
         """If title is not provided, it defaults to the provided rock name."""
-        if not title:
-            title = values.get("name", "")
-        return cast(str, title)
+        if not values.get("title"):
+            values["title"] = values.get("name")
+        return values
 
-    @pydantic.validator("parts", each_item=True)
+    @pydantic.field_validator("parts")
     @classmethod
     def _validate_base_and_overlay(
-        cls, item: dict[str, Any], values: dict[str, Any]
+        cls, parts: dict[str, Any], info: pydantic.ValidationInfo
     ) -> dict[str, Any]:
         """Projects with "bare" bases cannot use overlays."""
-        if values.get("base") == "bare" and part_has_overlay(item):
-            raise ValueError(
-                'Overlays cannot be used with "bare" bases (there is no system to overlay).'
-            )
-        return item
+        if info.data.get("base") != "bare":
+            return parts
 
-    @pydantic.validator("entrypoint_service")
+        for part in parts.values():
+            if part_has_overlay(part):
+                raise ValueError(
+                    'Overlays cannot be used with "bare" bases (there is no system to overlay).'
+                )
```

Comment:
It might be worth mentioning the specific part to the user:

```suggestion
        for name, part in parts.items():
            if part_has_overlay(part):
                raise ValueError(
                    f"Overlays cannot be used with "bare" bases (there is no system to overlay). (in part {name})"
                )
```

Or better, make this a functional validator on one part and make a `Part` annotated type that calls this validator.

Reply:
>  and make a Part annotated type that calls this validator

That was my original thought but I couldn't see how to do it without overriding the `parts` declaration in the craft-app Project. Maybe this validator should be upstreamed entirely?

Reply:
Added the part name as suggested, thanks!